tsweb/server.go

```diff
@@ -0,0 +1,385 @@
+	HTTP *http.Server
+
+	// Debug is the handler for /debug/ on HTTP and HTTPS.
+	Debug *DebugHandler
+
+	// AlwaysHeaders are HTTP headers that are set on all requests
+	// prior to invoking Handler.
+	AlwaysHeaders http.Header
+	// TLSHeaders are HTTP headers that are set on all TLS requests
+	// prior to invoking Handler.
+	TLSHeaders http.Header
+
+	// RequestLog is where HTTP request logs are written. If nil,
+	// request logging is disabled.
+	RequestLog func(*AccessLogRecord)
+
+	now func() time.Time // normally time.Now, modified for tests
+
+	vars         metrics.Set
+	httpRequests expvar.Int       // counter, completed requests
+	httpActive   expvar.Int       // gauge, currently alive requests
+	tlsRequests  metrics.LabelMap // counter, completed requests
+	tlsActive    metrics.LabelMap // gauge, currently alive requests
+	statusCode   metrics.LabelMap // status code of completed requests
+	statusFamily metrics.LabelMap // like statusCode, but bucketed by 1st digit of status code
+}
+
+// NewServer returns a Server, initialized by cfg with good defaults
+// for serving.
+func NewServer(cfg ServerConfig) *Server {
+	s := &Server{
+		Handler:       cfg.Handler,
+		Debug:         Debugger(http.NewServeMux()),
+		AlwaysHeaders: http.Header{},
+		TLSHeaders:    http.Header{},
+
+		tlsRequests:  metrics.LabelMap{Label: "version"},
+		tlsActive:    metrics.LabelMap{Label: "version"},
+		statusCode:   metrics.LabelMap{Label: "code"},
+		statusFamily: metrics.LabelMap{Label: "code_family"},
+	}
+
+	s.vars.Set("http_requests", &s.httpRequests)
+	s.vars.Set("gauge_http_active", &s.httpActive)
+	s.vars.Set("tls_request_version", &s.tlsRequests)
+	s.vars.Set("gauge_tls_active_version", &s.tlsActive)
+	s.vars.Set("http_status", &s.statusCode)
+	s.vars.Set("http_status_family", &s.statusFamily)
+
+	switch {
+	case cfg.ForceTLS:
+		s.HTTPS = &http.Server{
+			Addr:         cfg.Addr,
+			Handler:      s,
+			ReadTimeout:  defaultTimeout,
+			WriteTimeout: defaultTimeout,
+			IdleTimeout:  defaultTimeout,
+			TLSConfig: &tls.Config{
+				NextProtos: []string{
+					"h2", "http/1.1", // enable HTTP/2
+				},
+				MinVersion: tls.VersionTLS12,
+			},
+		}
+		// Deliberately don't forcibly enable HSTS here, this is a
+		// configuration where the caller has requested very manual
+		// TLS management, we shouldn't impose HSTS.
+	case IsProd443(cfg.Addr):
+		s.CertManager = &autocert.Manager{
+			Prompt:     autocert.AcceptTOS,
+			Cache:      autocert.DirCache(DefaultCertDir(cfg.Name)),
+			Email:      fmt.Sprintf("infra+autocert-%s@tailscale.com", cfg.Name),
+			HostPolicy: cfg.AllowedHostnames,
+		}
+		if s.CertManager.HostPolicy == nil {
+			s.CertManager.HostPolicy = func(context.Context, string) error {
+				return errors.New("no TLS serving allowed")
+			}
+		}
+		s.HTTPS = &http.Server{
+			Addr:         cfg.Addr,
+			Handler:      s,
+			ReadTimeout:  defaultTimeout,
+			WriteTimeout: defaultTimeout,
+			IdleTimeout:  defaultTimeout,
```

Comment:
10 seconds idle timeout is quite a bit lower than what we currently have in places, isn't it?